scene/resources/dynamic_font.cpp

```diff
@@ -303,14 +328,182 @@ void DynamicFontAtSize::set_texture_flags(uint32_t p_flags) {
+					}
+				} else {
+					if (current_cluster != NULL) {
+						new_cluster->max_index = current_cluster->min_index - 1;
+					} else {
+						new_cluster->max_index = p_start + p_length - 1;
+					}
+				}
+
+				current_cluster = new_cluster;
+				last_cluster_id = glyphInfo[i].cluster;
+			} else {
+				current_cluster->glyphs.push_back(GlyphInfo(glyphInfo[i].codepoint, Point2((glyphPos[i].x_offset) / 64, -(glyphPos[i].y_offset / 64)), Point2((glyphPos[i].x_advance) / 64, (glyphPos[i].y_advance / 64))));
+				current_cluster->valid &= ((glyphInfo[i].codepoint != 0) || !u_isgraph(p_utf16_text[glyphInfo[i].cluster])); //have valid glyph index or non graphic char
+				current_cluster->length += glyphPos[i].x_advance / 64;
+
+				if (p_index == -1) {
+					valid &= current_cluster->valid;
+					offset_x += glyphPos[i].x_advance / 64;
+				} else {
+					//fallback
+					if (!p_visual_order[p_pre_shape_clusters + cluster_count]->valid) {
+						valid &= current_cluster->valid;
+						offset_x += glyphPos[i].x_advance / 64;
+					} else {
+						//fix cluster offset/size
+						offset_x += p_visual_order[p_pre_shape_clusters + cluster_count]->glyphs[current_cluster->glyphs.size() - 1].advance.x;
+					}
+				}
+			}
+		}
+		if (p_direction == DIR_LTR) {
+			current_cluster->max_index = p_start + p_length - 1;
+		}
+		//DEBUG
+		//printf("RUN_SHAPE: clusters_processed: %d, clusters replaced: %d\n", cluster_count, fallback_replaced_cluster_count);
+	}
+
+	hb_buffer_destroy(buffer);
+
+	return valid;
+}
+
+void DynamicFontAtSize::draw_cluster(RID p_canvas_item, const Point2 &p_pos, const Ref<ClusterInfo> &p_cluster, const Color &p_modulate, const Vector<Ref<DynamicFontAtSize> > &p_fallbacks) const {
+	if (p_cluster == NULL)
+		return;
+
+	if (p_cluster->fallback_index < 0) {
+		Point2 offset = Point2(p_cluster->get_offset(), 0);
+		for (int i = 0; i < p_cluster->glyphs.size(); i++) {
+			offset += p_cluster->glyphs[i].offset;
+			draw_raw_glyph(p_canvas_item, p_pos + offset, p_cluster->glyphs[i].codepoint, p_modulate);
+		}
+	} else {
+		if (p_cluster->fallback_index < p_fallbacks.size()) {
+			DynamicFontAtSize *fb = const_cast<DynamicFontAtSize *>(p_fallbacks[p_cluster->fallback_index].ptr());
+			if (!fb->valid)
+				return;
+
+			Point2 offset = Point2(p_cluster->get_offset(), 0);
+			for (int i = 0; i < p_cluster->glyphs.size(); i++) {
+				offset += p_cluster->glyphs[i].offset;
+				fb->draw_raw_glyph(p_canvas_item, p_pos + offset, p_cluster->glyphs[i].codepoint, p_modulate);
+			}
+		}
+	}
+}
+
+void DynamicFontAtSize::draw_raw_glyph(RID p_canvas_item, const Point2 &p_pos, uint32_t p_glyph, const Color &p_modulate) const {
+	if (!valid)
+		return;
+
+	const_cast<DynamicFontAtSize *>(this)->_update_glyph(p_glyph);
+
+	const Character *c = glyph_map.getptr(p_glyph);
+
+	if ((!c->found) || (p_glyph == 0)) {
+		//not found, try 0xFFFD to display 'not found'.
+
+		const_cast<DynamicFontAtSize *>(this)->_update_char(0xFFFD);
```

Comment:
0xFFFD usually is used in case of an encoding error to indicate invalid characters. For characters missing from the font, glyph id 0 is usually used (in most fonts this is an empty rectangle) or if you want some thing fancy draw a hex box:
![](http://www.adambaker.org/unicode_input_screenshot.png)